### ISSUE TITLE
docs: add eval datasets — load JSONL task sets + export passing runs as SFT data (PraisonAI PR #5000)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#4064]
## Context

PraisonAI PR [#5000](https://github.com/MervinPraison/PraisonAI/pull/5000) (merged 2026-09-08, branch `fix/eval-jsonl-and-sft` → `main`) added a new **eval datasets** module that connects two things that were sitting in the repo unconnected:

1. `EvalCase.from_dict` existed but **nothing read a file** — a task set had to be assembled in Python.
2. `praisonai-train` had no link back to `eval` — no way to take the runs that scored well and turn them into fine-tuning data.

The PR adds five new public helpers under `praisonaiagents.eval` — `load_cases`, `iter_jsonl`, `export_sft`, `sft_records`, and `DatasetError` — and none of them are documented today. `grep`-ing the docs repo for `load_cases`, `export_sft`, or `datasets` returns zero hits. The existing [`docs/features/eval-suite.mdx`](https://github.com/MervinPraison/PraisonAIDocs/blob/main/docs/features/eval-suite.mdx) covers running the evaluators but not loading task sets from a file or exporting passing runs.

**Merged PR (source of truth):** https://github.com/MervinPraison/PraisonAI/pull/5000
**Head SHA:** `48c4ae207cb6e643361847bbaefb6ae54c9dce9f`

## What the feature does — for the docs writer

**Goal:** close the loop from **task set → run → measure → train**, using JSONL at both ends (the format Agno, OpenAI's fine-tuning API, and every dataset tool already speak).

```
tasks.jsonl  ──► load_cases  ──► suite.run(cases)  ──► EvalReport  ──► export_sft  ──► train.jsonl
                                                                                       │
                                                                                       ▼
                                                                       praisonai-train llm --dataset train.jsonl
```

**One-liner from the PR body:**

```python
cases  = load_cases("tasks.jsonl")
report = suite.run(cases)
export_sft(report, "train.jsonl", min_score=0.8)
```

### Design principles (surface these in the docs, verbatim if possible)

**Forgiving where it costs nothing:**
- Common column aliases are accepted — `prompt` / `question` / `query` / `instruction` all map to `input`; `expected_output` / `answer` / `output` / `completion` all map to `expected`. A task set exported from another tool is the normal case, not the exception.
- Unrecognised columns are **kept as metadata** rather than dropped. Datasets usually carry provenance (`source`, `difficulty`, `id`) worth having in the report.
- Blank lines and `//` comment lines are skipped.
- Missing `name` field auto-populates as `case_1`, `case_2`, … (by position among data rows).

**Strict where it matters:**
- A malformed line is reported by `file:line` — "invalid JSON" is not an actionable message on a thousand-line dataset.
- The physical file line number is preserved past skipped blank/comment lines, so `file:900` really is line 900 in the editor.
- A non-object line (e.g. `[1, 2, 3]`) is refused with "must be a JSON object".
- An empty file **raises** `DatasetError("no cases")` — an empty task set would otherwise report a perfect pass rate over nothing, "the most flattering possible lie an eval can tell".

**Export refuses to lie the same way:**
- Only **passing** runs at or above `min_score` are exported. Training on a run the evaluation *rejected* teaches the model the exact behaviour the eval exists to catch.
- An export with nothing to write **raises**, rather than producing an empty file that would look like a successful export.
- A run missing either a prompt or a completion is skipped silently would quietly shrink the training set — the module refuses instead.

## SDK surface

**Imports (friendly, top-level — matches the AGENTS.md "no complex imports" rule):**

```python
from praisonaiagents.eval import (
    load_cases,
    iter_jsonl,
    export_sft,
    sft_records,
    DatasetError,
)
```

All five are lazy-loaded in `praisonaiagents/eval/__init__.py` (both `__all__` and the `_LAZY_IMPORTS` map).

### `load_cases(path, *, case_cls=None) -> List[EvalCase]`

Reads a JSONL task set into `EvalCase` objects. Returns a list of cases with the aliasing / metadata / auto-naming rules described above.

| Param | Type | Default | Description |
|-------|------|---------|-------------|
| `path` | `str` | required | Path to the `.jsonl` file |
| `case_cls` | `type` | `EvalCase` | Override the case class (advanced) |

**Raises:** `DatasetError` on missing file, malformed line, non-object line, or empty file — always with `file:line` when the fault is inside the file.

### `iter_jsonl(path) -> Iterator[Dict[str, Any]]`

Yields one dict per data line. Same error handling as `load_cases`, but no aliasing / metadata folding. Useful for arbitrary JSONL data (not just eval cases).

### `export_sft(report, path, *, min_score=1.0, system_prompt=None) -> int`

Writes passing runs to `path` as JSONL. Returns the number of records written.

| Param | Type | Default | Description |
|-------|------|---------|-------------|
| `report` | `EvalReport` or `List[EvalResult]` | required | Anything exposing `.results`, or a bare list |
| `path` | `str` | required | Destination `.jsonl` (parent dirs are created) |
| `min_score` | `float` | `1.0` | Only runs with `score >= min_score` are exported |
| `system_prompt` | `str \| None` | `None` | Prepended as a `system` message when set |

**Raises:** `DatasetError` if no run scored high enough, or if `report` is not an `EvalReport` / list.

### `sft_records(report, *, min_score=1.0, system_prompt=None) -> List[Dict[str, Any]]`

Same selection logic as `export_sft` but returns the records in memory instead of writing them. Useful when the caller wants to post-process or merge multiple exports before writing.

### `DatasetError(ValueError)`

Raised for every dataset read / write failure. Subclass of `ValueError`, so existing `except ValueError` handlers still catch it.

## Output format (SFT chat records)

One JSON object per line, chat-format (the shape OpenAI's fine-tuning API expects):

```jsonl
{"messages":[{"role":"user","content":"2+2?"},{"role":"assistant","content":"4"}]}
```

With a `system_prompt`:

```jsonl
{"messages":[{"role":"system","content":"Be terse."},{"role":"user","content":"2+2?"},{"role":"assistant","content":"4"}]}
```

The prompt is read from `result.record["input"]`; the completion is `result.actual_output`.

## Input JSONL format (task sets)

Minimal:

```jsonl
{"name": "add", "input": "2+2?", "expected": "4"}
{"name": "capital", "input": "capital of France?", "expected": "Paris"}
```

With aliases + extra columns folded into metadata:

```jsonl
{"prompt": "2+2?", "answer": "4", "difficulty": "easy", "source": "sheet1"}
```

Blank lines and `//` comments allowed:

```jsonl
// Arithmetic set
{"input": "2+2?", "expected": "4"}

{"input": "3+3?", "expected": "6"}
```

## Verification (from the merged PR)

- **15 new tests, all passing** (`tests/unit/eval/test_datasets.py`) — including two controls (a weak pass exports when the bar is lower; blank/comment lines are skipped rather than failing the file).
- Whole `tests/unit/eval` suite: **389 passed** on this branch vs **374 passed** on `origin/main`, exactly **+15** with zero regressions.
- All parity gates (`names` / `signatures` / `behaviour`) green.

---

## Documentation changes required

All new content goes in `docs/features/*` per the AGENTS.md placement rules (**never** `docs/concepts/`). Follow the AGENTS.md style guide — hero mermaid diagram at the top, `<Steps>` / `<AccordionGroup>` / `<CardGroup>` components, standard colour scheme (`#8B0000`, `#189AB4`, `#10B981`, `#F59E0B`, `#6366F1`, white text, `#7C90A0` stroke), one-sentence section intros, no forbidden phrases.

### 1. New page: `docs/features/eval-datasets.mdx`

Create a new feature page dedicated to loading task sets from JSONL and exporting passing runs as SFT data. Suggested structure:

- **Frontmatter:**
  ```yaml
  ---
  title: "Eval Datasets"
  sidebarTitle: "Eval Datasets"
  description: "Load JSONL task sets and export passing runs as fine-tuning data"
  icon: "database"
  ---
  ```
- **Hero blurb + mermaid diagram** — one sentence + a `graph LR` showing `tasks.jsonl → load_cases → EvalSuite → EvalReport → export_sft → train.jsonl`. Use the standard colour scheme (Input=indigo, process=teal, output=green).
- **Quick Start (`<Steps>`)** — three steps:
  1. **Load a task set** — the minimal `load_cases("tasks.jsonl")` example.
  2. **Run the suite on the loaded cases** — pass `cases` into an `EvalSuite` (or a runner) and get an `EvalReport`.
  3. **Export what passed** — `export_sft(report, "train.jsonl", min_score=0.8)`.
- **Input format** — the minimal JSONL block, the aliases-and-metadata block, and the blank-line/comment block from above, each with a one-sentence intro.
- **Output format (SFT)** — the `messages` chat-format block, and the variant with a `system_prompt`.
- **Behaviour worth knowing** — bulleted:
  - Alias table: `input` ← `prompt` / `question` / `query` / `instruction`; `expected` ← `expected_output` / `answer` / `output` / `completion`.
  - Unrecognised columns fold into `metadata` (not dropped).
  - Auto-naming: `case_1`, `case_2`, … by position among data rows.
  - Blank + `//` comment lines skipped; physical file line preserved in errors.
  - Empty file raises `DatasetError` — no silent perfect-score.
  - Only `passed=True` runs with `score >= min_score` are exported.
  - Zero-record export raises `DatasetError` — no silent empty file.
  - `DatasetError` subclasses `ValueError`, so existing `except ValueError` still catches it.
- **Python API** — signatures and parameter tables for all five public names (`load_cases`, `iter_jsonl`, `export_sft`, `sft_records`, `DatasetError`) using the tables above.
- **Common Patterns** section — 2–3 short recipes:
  - **Portable in, portable out** — one snippet that loads a Hugging Face-exported JSONL (with `prompt` / `answer` columns) and exports a fine-tuning file OpenAI's API accepts as-is.
  - **Lower the bar for a small dataset** — set `min_score=0.5` to keep more attempts when the suite is small; note this is a deliberate trade against SFT quality.
  - **Persist a system prompt with each training row** — use the `system_prompt=` kwarg so the fine-tuned model learns the same persona the eval was scored under.
- **Best Practices (`<AccordionGroup>`)** — 3–4 accordions:
  - *"Fold provenance columns into metadata, don't strip them"* — the loader keeps `source` / `difficulty` / `id` automatically; you'll want them in the report.
  - *"Set `min_score` above your CI gate, not equal to it"* — a run that barely passed is a fragile teacher.
  - *"Never edit a training file by hand"* — regenerate from the eval report so provenance stays honest.
  - *"Keep `//` comment lines in the source dataset"* — they're skipped at load time and make the file self-documenting.
- **Callout (`<Warning>`)** — the honest-selection caveat, verbatim from the PR:

  > `min_score`-filtered export is *rejection sampling* — it amplifies behaviour the agent already produces and inherits any bias in the judge that decided a run "passed". It cannot teach behaviour the agent never exhibited; treat it as reinforcing verified wins, not as an oracle.

- **Related (`<CardGroup>`)** — link to `eval-suite`, `accuracy-evaluator`, `praisonai-train-dataset-tooling`, `praisonai-train-package`, and (once documented) `run-fingerprint` (PR #4971 — the PR body calls out that this pair covers "load a task set, run it, know the setup didn't move, export what worked").

### 2. Update `docs/features/eval-suite.mdx`

- In the **Related** `<CardGroup>` at the bottom, add a new `<Card>`:
  ```
  <Card title="Eval Datasets" icon="database" href="/features/eval-datasets">
    Load task sets from JSONL and export passing runs as SFT data
  </Card>
  ```
- Optionally add a one-line note near Quick Start explaining that `suite.run(cases)` accepts the list of `EvalCase` objects that `load_cases` returns.

### 3. Update `docs.json`

- Add `"features/eval-datasets"` to the **Features** group in `docs.json` (near the other eval pages: `eval-suite`, `context-evaluator`, `loop-evaluator`, `harness-evaluator`).
- **Do not** add anything under the **Concepts** group — AGENTS.md forbids AI agents from touching `docs/concepts/*`.
- Verify `docs.json` remains valid JSON after the edit.

### 4. (If it exists) update `docs/features/praisonai-train-package.mdx` / `docs/train.mdx`

Add a one-line "See also" pointer from any page that documents fine-tuning inputs, since `export_sft` is now the recommended way to build a trainer-ready dataset from a scored eval run. Example line:

> **From eval runs:** [`export_sft`](/features/eval-datasets#export_sft) turns the passing runs of an `EvalReport` into a chat-format JSONL that this trainer consumes unchanged.

## SDK verification checklist for the writer

Before writing, `git pull` the PraisonAI repo and read each of these files in full (SDK-first per AGENTS.md — never document what isn't in the source):

- [ ] `src/praisonai-agents/praisonaiagents/eval/datasets.py` — **the whole file (184 lines)**; author every signature, default, and behaviour statement from this file, not from this issue.
- [ ] `src/praisonai-agents/praisonaiagents/eval/__init__.py` — confirm the five names in `__all__` and `_LAZY_IMPORTS` are the same names to document.
- [ ] `src/praisonai-agents/praisonaiagents/eval/package.py` — read the `EvalCase` and `EvalResult` dataclasses to describe the loader's target shape (`name` / `input` / `expected` / `criteria` / `metadata` / `timeout_seconds` on `EvalCase`; `case_name` / `passed` / `score` / `actual_output` / `record` on `EvalResult`).
- [ ] `src/praisonai-agents/tests/unit/eval/test_datasets.py` — **137 lines, 15 tests**; every documented behaviour must correspond to a test here (e.g. `test_common_column_aliases_are_accepted`, `test_unrecognised_columns_are_kept_as_metadata`, `test_error_reports_physical_line_past_skipped_lines`, `test_control_a_weak_pass_exports_when_the_bar_is_lower`).
- [ ] Existing `docs/features/eval-suite.mdx` — for house style and where to slot the "Eval Datasets" card.

## Placement & branch reminder (per AGENTS.md)

- All new files/edits go in `docs/features/*` — **never** `docs/concepts/*`.
- Update `docs.json` **only** under the "Features" group.
- Create the change on a feature branch (not `main`) and open a **draft** PR.
- Use Mintlify components (`Steps`, `AccordionGroup`, `CardGroup`) and the standard mermaid colour scheme.

---

_Filed by the docs-sync scheduled task after reviewing PraisonAI PR [#5000](https://github.com/MervinPraison/PraisonAI/pull/5000) (merged 2026-09-08). Ready for an implementation agent to pick up._
